# Pandas fundamentals

#### Written for the QuantEcon Indian Summer Workshop (August 2022)
#### Author: [Shu Hu](https://shu-hu.com/intro.html)

In [1]:
import pandas as pd

Given the dataframe below you are required to do the following exercises.

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/QuantEcon/lecture-python-programming/master/source/_static/lecture_specific/pandas/data/test_pwt.csv')

In [3]:
df

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
3  64.436451  10.266688  
4  74.707624  11.658954  
5  72.718710   5.726546  
6  72.347054   6.032454  
7  78.978740   5.108068

### Exercise 1 (View data)

### Exercise 1.1

Show the top 5 rows of the dataframe ``df``.

### Exercise 1.2

Show the bottom 5 rows of the dataframe ``df``.

### Exercise 1.3

Show a quick statistic summary of the dataframe ``df``.

### Exercise 1.4

Transpose the dataframe ``df``.

### Exercise 1.5

Sort the dataframe ``df`` by values in column ``tcgdp`` ascendingly.

### Exercise 2 (Select data)

### Exercise 2.1

Select the column ``tcgdp`` from the dataframe ``df``, yielding a series.

### Exercise 2.2

Select the top 3 rows from the dataframe ``df``.

### Exercise 2.3


Get the first row of data across all the columns.

### Exercise 2.4

Select all values on two columns ``country`` and ``POP``.

### Exercise 2.5

Select values in rows 2-4 on two columns ``country`` and ``POP``.

### Exercise 2.6

Get a scalar value for the 2nd row in the column ``POP`` which should be ``1006300.297``.

### Exercise 3 (Boolean indexing)

### Exercise 3.1 

Select data with column ``POP``'s values greater than ``30_000``.

### Exercise 3.2

We see

In [19]:
df1 = df.copy()
df1

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  
0  75.716805   5.578804  
1  67.759026   6.720098  
2  64.575551  14.072206  
3  64.436451  10.266688  
4  74.707624  11.658954  
5  72.718710   5.726546  
6  72.347054   6.032454  
7  78.978740   5.108068

In [20]:
df1["X"] = ["A", "B", "C", "D", "E", "F", "G","H"]

In [21]:
df1

country country isocode  year          POP       XRAT         tcgdp  \
0      Argentina             ARG  2000    37335.653   0.999500  2.950722e+05   
1      Australia             AUS  2000    19053.186   1.724830  5.418047e+05   
2          India             IND  2000  1006300.297  44.941600  1.728144e+06   
3         Israel             ISR  2000     6114.570   4.077330  1.292539e+05   
4         Malawi             MWI  2000    11801.505  59.543808  5.026222e+03   
5   South Africa             ZAF  2000    45064.098   6.939830  2.272424e+05   
6  United States             USA  2000   282171.957   1.000000  9.898700e+06   
7        Uruguay             URY  2000     3219.793  12.099592  2.525596e+04   

          cc         cg  X  
0  75.716805   5.578804  A  
1  67.759026   6.720098  B  
2  64.575551  14.072206  C  
3  64.436451  10.266688  D  
4  74.707624  11.658954  E  
5  72.718710   5.726546  F  
6  72.347054   6.032454  G  
7  78.978740   5.108068  H

Filter rows containing ``A``, ``C``, ``G`` in column ``X`` using method [isin()](https://pandas.pydata.org/docs/reference/api/pandas.Series.isin.html#pandas.Series.isin).